File: 백준/Gold/4179. 불！/불！.c
#include <stdio.h>
#include <stdbool.h>
#include <memory.h>
#define BUFSIZE 1 << 18
char readbuf[BUFSIZE];
int rp = BUFSIZE;
char ReadChar(){
    if(rp == BUFSIZE){
        fread(readbuf, 1, BUFSIZE, stdin);
        rp = 0;
    }
    return readbuf[rp++];
}
int ReadInt(){
    char c;
    int ret = 0;
    while(c < '0' || c > '9') c = ReadChar();
    while(c >= '0' && c <= '9'){
        ret = ret * 10 + (c & 0xf);
        c = ReadChar();
    }
    return ret;
}
#define SIZE 100000
typedef struct{
    int x, y;
}Point;
Point queue[SIZE];
int front = -1, rear = -1;
int dx[4] = {0, 0, -1, 1};
int dy[4] = {-1, 1, 0 ,0};
int visit[1001][1001];
int visitFire[1001][1001];
char map[1001][1001];
int R, C;

void push(Point data){
    rear = (rear + 1) % SIZE;
    queue[rear] = data;
}
Point pop(){
    front = (front + 1) % SIZE;
    return queue[front];
}

void bfsFire(){

    while(front != rear){
        Point cur = pop();
        for(int i = 0; i < 4; i++){
            int nx = cur.x + dx[i];
            int ny = cur.y + dy[i];
            if(nx < 0 || ny < 0 || nx > R - 1 || ny > C - 1) continue;
            if(visitFire[nx][ny] || map[nx][ny] == '#') continue;
            Point new = {nx, ny};
            push(new);
            visitFire[nx][ny] = visitFire[cur.x][cur.y] + 1;
        }
    }
}           
int bfs(Point src){
    push(src);
    while(front != rear){
        Point cur = pop();
        for(int i = 0; i < 4; i++){
            int nx = cur.x + dx[i];
            int ny = cur.y + dy[i];
            if(nx < 0 || ny < 0 || nx > R - 1 || ny > C - 1){
                return visit[cur.x][cur.y];
            }
            if(visit[nx][ny] || map[nx][ny] == '#') continue;
            if(visitFire[nx][ny] && visitFire[nx][ny] <= visit[cur.x][cur.y] + 1) continue;
            Point new = {nx, ny};
            push(new);
            visit[nx][ny] = visit[cur.x][cur.y] + 1;
        }
    }
    return -1;
}

int main(void){
    Point src;
    R = ReadInt(), C = ReadInt();
    for(int i = 0; i < R; i++){
        for(int j = 0; j < C; j++){
            map[i][j] = ReadChar();
            if(map[i][j] == 'F'){
                Point srcFire;
                srcFire.x = i;
                srcFire.y = j;
                push(srcFire);
                visitFire[i][j] = 1;
            }else if(map[i][j] == 'J'){
                src.x = i;
                src.y = j;
                visit[i][j] = 1;
            }
        }
        ReadChar();
    }
    bfsFire();
    int ans = bfs(src);
    if(ans == -1) printf("IMPOSSIBLE\n");
    else printf("%d\n", ans);
}

File: 백준/Gold/4179. 불！/불！.py
import collections
import sys
input = sys.stdin.readline

dx = (0, 0, -1, 1)
dy = (-1, 1, 0, 0)

def bfsFire():
    while fQ:
        x, y = fQ.popleft()
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            if not (0 <= nx < R and 0 <= ny < C): continue
            if visitFire[nx][ny] or map[nx][ny] == '#': continue
            visitFire[nx][ny] = visitFire[x][y] + 1
            fQ.append([nx, ny])

def bfs():
    while hQ:
        x, y = hQ.popleft()
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            if not (0 <= nx < R and 0 <= ny < C):
                print(visit[x][y])
                return
            if visit[nx][ny] or map[nx][ny] == '#': continue
            if visitFire[nx][ny] and visitFire[nx][ny] <= visit[x][y] + 1: continue
            visit[nx][ny] = visit[x][y] + 1
            hQ.append([nx, ny])
    print("IMPOSSIBLE")
    return


R, C = map(int, input().split())
visit = [[0 for _ in range(C)] for _ in range(R)]
visitFire = [[0 for _ in range(C)] for _ in range(R)]
map = []

fQ = collections.deque()
hQ = collections.deque() 

for i in range(R):
    map.append(list(input().strip()))

for i in range(R):
    for j in range(C):
        if map[i][j] == 'F':
            fQ.append([i, j])
            visitFire[i][j] = 1
        elif map[i][j] == 'J':
            hQ.append([i, j])
            visit[i][j] = 1
bfsFire()
bfs()
